### PR TITLE
test(ui): Add Cypress tests for workload cve image single page

### PR DESCRIPTION
**Pull request by dvail**
## Description

Additional Cypress tests for Workload CVEs **Image single** page.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

CI e2e tests

### CODE REVIEW

**Comment by dvail** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6633** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6633" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6633?utm_source=stack-comment).

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        // Check that only applicable resource menu items are present in the toolbar
+        cy.get(selectors.resourceDropdown).click();
+        cy.get(selectors.resourceMenuItem('CVE'));
+        cy.get(selectors.resourceMenuItem('Image')).should('not.exist');
+        cy.get(selectors.resourceMenuItem('Deployment')).should('not.exist');
+        cy.get(selectors.resourceMenuItem('Cluster')).should('not.exist');
+        cy.get(selectors.resourceMenuItem('Namespace')).should('not.exist');
+        cy.get(selectors.resourceDropdown).click();
+
+        // Check that the CVEs by severity totals in the card match the number in the "results found" text
```
Comment (pedrottimark) [inline ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:31]: What do you think about separating into 2 tests:
1. preceding assertions are independent of data
2. following sets of assertions are dependent on data (keep together to ensure that same page is tested)

Although computation looks reliable, separation would allow a temporary `skip` in worst case if we discovered later that it failed for edge case of no results.
No_results_found

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            cy.get(`*:contains(${statusTotal} result${plural} found)`);
+        });
+    });
+
+    it('should correctly apply severity filters', () => {
+        visitWorkloadCveOverview();
+
+        // Go to the image tab
+        cy.get(selectors.entityTypeToggleItem('Image')).click();
+
```
Comment (pedrottimark) [inline ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:71]: Although pro and con of selectors in test file versus encapsulated in helpers file, comment and multiple calls is some evidence in favor of a semantically named helper function with `'Image'` as argument.

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        // Select any image that has CVEs in the table
+        cy.get(
+            `tbody tr:has(td[data-label="CVEs by severity"] ${selectors.nonZeroCveSeverityCounts})`
+        )
+            .first()
+            .then(([$rowWithCves]) => {
+                // Click the link in the table to visit the image page
+                cy.wrap($rowWithCves).find('td[data-label="Image"] a').click();
```
Comment (pedrottimark) [inline ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:156]: In this test, image page does not depend on any specific information from images page (bravo).

Use of `then` suggests that assertions for second page depend on information from first page, as in the previous test.

To flatten the test, what do you think about either of the following with method or pseudo-selector:

```js
cy.get(
    `tbody tr:has(td[data-label="CVEs by severity"] ${selectors.nonZeroCveSeverityCounts})`
).first().find('td[data-label="Image"] a').click();
```

```js
cy.get(`tbody tr:has(td[data-label="CVEs by severity"] ${selectors.nonZeroCveSeverityCounts}):first td[data-label="Image"] a`).click();
```

P.S. Your discovery of `has` pseudo-selector opens the door to this possibility. Earlier tests needed `then` block to do this.

**Review by pedrottimark — APPROVED** [review]
Super use of accessibility and `data-label` instead of `data-testid` attributes in tests.

Thank you for continued follow up. A few comments for your consideration.

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import withAuth from '../../../helpers/basicAuth';
+import {
+    applyLocalSeverityFilters,
+    selectResourceFilterType,
+    typeAndEnterResourceFilterValue,
+    visitWorkloadCveOverview,
+} from './WorkloadCves.helpers';
+import { selectors } from './WorkloadCves.selectors';
+
+describe('Image Single page', () => {
```
[the remark this answers is not in the document]
Comment (dvail) [reply ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:12]: I like this, seeing `Image Single page` in a test failure would be pretty vague.

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js (hunk 3 above)
Comment (dvail) [reply ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:156]: Great catch, I wasn't thrilled about the `then` block here but didn't think to use `find`!

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
+import capitalize from 'lodash/capitalize';
```
Comment (pedrottimark) [inline ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:1]: Forgot to mention a subtle distinction that caused problems in VM 1.0 for image CVE or node CVE
* https://lodash.com/docs/4.17.15#capitalize
    > Converts the first character of string to upper case **and the remaining** to lower case.
* https://lodash.com/docs/4.17.15#upperFirst
    > Converts the first character of string to upper case.

File: ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js (hunk 3 above)
Comment (dvail) [reply ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js:156]: Style-wise I liked the latter, but it looks like it isn't possible to chain the `:has()` and `:first` pseudo-selectors. I checked the jquery [docs](https://api.jquery.com/first-selector/) and it looks like `:first` is deprecated anyway and usage of `.first()` is preferred.

Comment (pedrottimark) [reply ui/apps/platform/cypress/integration/vulnerabilities/workloadCves/imageSingle.test.js]: Aha, what I meant when I wrote `:first` was `:nth(0)` although not necessarily recommending it.

**Comment by pedrottimark** [comment]
Dave, although optional to retest: ocp-4-10-ui-e2e-tests are passing again after change from fast to stable channel

**Comment by dvail** [comment]
Thanks Mark, I actually think this might have been an issue with my test and pushed a fix. I've been trying to make sure all jobs are green lately even if I'm pretty sure the failure is unrelated just to be safe. 😅
